Jenkinsfile

```diff
@@ -1,67 +1,145 @@
+                         condaInstallDevito()
+                         installYask()
+                         runCondaTests()
+                     }
+                }
             }
         }
     }
-}               
+}
 
-// Scripted pipeline method for building and testing a docker container with devito
-def buildAndTest (def gccvers, def DEVITO_BACKEND=null, def DEVITO_OPENMP=0, def OMP_NUM_THREADS=1) {
-    // Switch into scripted pipeline
-    script {
-        // If we're specifying a backend, pass it through, and label the container appropriately
-        BACKEND_ARG = DEVITO_BACKEND ? "--build-arg DEVITO_BACKEND=${DEVITO_BACKEND}" : ''
-        BACKEND_SUFFIX = DEVITO_BACKEND ? "-${DEVITO_BACKEND}" : ''
-        OPENMP_SUFFIX = (DEVITO_OPENMP=='1') ? "-openmp" : ''
-        /* Now build the container, saving the returned image value to run with testing
-           Note that the context directory has to be the final parameter passed */
-        def customImage = 
-             docker.build("opesci/devito-jenkins:gcc${gccvers}${BACKEND_SUFFIX}${OPENMP_SUFFIX}-${env.BUILD_ID}", 
-                         "--no-cache -f Dockerfile.jenkins --build-arg gccvers=${gccvers} ${BACKEND_ARG} .")
-        // If the build succeeded, push the container to dockerhub for debugging purposes
-        customImage.push()
-        // Also push a latest tagged version for easy reference
-        customImage.push('gcc${gccvers}${BACKEND_SUFFIX}${OPENMP_SUFFIX}-latest')
-        // Using the built container, run tests
-        customImage.inside("-e DEVITO_OPENMP=${DEVITO_OPENMP} -e OMP_NUM_THREADS=${OMP_NUM_THREADS}") {
-            sh "flake8 --builtins=ArgumentError ."
-            sh "py.test -n 2 -vs --cov"
-            // Additional seismic operator tests
-            if ( DEVITO_BACKEND!='yask' ) {
-                sh "python examples/seismic/benchmark.py test -P tti -so 4 -a -d 20 20 20 -n 5"
-                sh "python examples/seismic/benchmark.py test -P acoustic -a"
-                sh "python examples/seismic/acoustic/acoustic_example.py --full"
-                sh "python examples/seismic/acoustic/acoustic_example.py --constant --full"
-                sh "python examples/seismic/acoustic/gradient_example.py"
-                sh "python examples/misc/linalg.py mat-vec mat-mat-sum transpose-mat-vec"
-                sh "python examples/seismic/tti/tti_example.py -a"
-                sh "python examples/checkpointing/checkpointing_example.py"
-                // Test tutorial notebooks for the website using nbval
-                sh "py.test -vs --nbval examples/seismic/tutorials"
-                sh "py.test -vs --nbval examples/cfd"
-            }
-            // Now run coverage
-            sh "codecov"
-            // Now build documentation
-            sh "sphinx-apidoc -f -o docs/ examples/"
-            sh "sphinx-apidoc -f -o docs/ devito/ devito/yask/*"
-            sh "make -C docs/ html"
-        }
+def cleanWorkspace() {
```

Comment:
dumb question: why do we need ``cleanWorkspace``? don't we get a new, clean VM for each build?

Reply:
Frustratingly, this seems to be a cost of agent persistence on azure. I've set the agents to spin up and then remain active for (I think...) two hours so if there's a continuous stream of builds going on it avoids having to wait for the spin-up time on subsequent builds. However, Jenkins seems not to clean up the working source on the agent, at least in terms of removing extra files which are put in on top of the git checkout, ie. the conda environment install which in this case happens inside the source directory as it's the only place guaranteed to be writeable by the jenkins user on the agent. I think this is probably a bug in Jenkins which I should report, but as a fix for now I put in the cleanup stage. Even git clean -f -d isn't enough as that doesn't touch the yask repo which is nested inside the devito checkout...

Reply:
ok, that makes sense, thanks